planner/core/optimizer.go

```diff
@@ -37,6 +37,7 @@ const (
 	flagDecorrelate
 	flagEliminateAgg
 	flagEliminateProjection
+	flagEliminateOuterJoin
```

Comment:
In order to eliminate the following case, maybe we can make this optimization happens after predicate push down.

```sql
TiDB(localhost:4000) > desc select t1.a from t1 left outer join t2 on t1.a=t2.a where t1.b > 10;
+----------------------------+----------+------+--------------------------------------------------------------------------------+
| id                         | count    | task | operator info                                                                  |
+----------------------------+----------+------+--------------------------------------------------------------------------------+
| Projection_6               | 4166.67  | root | test.t1.a                                                                      |
| └─IndexJoin_9              | 4166.67  | root | left outer join, inner:TableReader_8, outer key:test.t1.a, inner key:test.t2.a |
|   ├─TableReader_13         | 3333.33  | root | data:Selection_12                                                              |
|   │ └─Selection_12         | 3333.33  | cop  | gt(test.t1.b, 10)                                                              |
|   │   └─TableScan_11       | 10000.00 | cop  | table:t1, range:[-inf,+inf], keep order:false, stats:pseudo                    |
|   └─TableReader_8          | 10.00    | root | data:TableScan_7                                                               |
|     └─TableScan_7          | 10.00    | cop  | table:t2, range: decided by [test.t1.a], keep order:false, stats:pseudo        |
+----------------------------+----------+------+--------------------------------------------------------------------------------+
7 rows in set (0.00 sec)

TiDB(localhost:4000) > desc select t1.a from t1 left outer join t2 on t1.a=t2.a where t1.a > 10;
+----------------------------+----------+------+-------------------------------------------------------------------------+
| id                         | count    | task | operator info                                                           |
+----------------------------+----------+------+-------------------------------------------------------------------------+
| Projection_6               | 4166.67  | root | test.t1.a                                                               |
| └─HashLeftJoin_11          | 4166.67  | root | left outer join, inner:TableReader_16, equal:[eq(test.t1.a, test.t2.a)] |
|   ├─TableReader_14         | 3333.33  | root | data:Selection_13                                                       |
|   │ └─Selection_13         | 3333.33  | cop  | gt(test.t1.a, 10)                                                       |
|   │   └─TableScan_12       | 10000.00 | cop  | table:t1, range:[-inf,+inf], keep order:false, stats:pseudo             |
|   └─TableReader_16         | 3333.33  | root | data:TableScan_15                                                       |
|     └─TableScan_15         | 3333.33  | cop  | table:t2, range:(10,+inf], keep order:false, stats:pseudo               |
+----------------------------+----------+------+-------------------------------------------------------------------------+
7 rows in set (0.00 sec)
```